api: Support user_id in get_user_presence endpoint.

tushar912 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
Fixes #14304

**Testing plan:** <!-- How have you tested? -->
tested manually as well as using ./tools/test-backend

timabbott commented on zerver/openapi/openapi.py:
```diff
@@ -126,11 +126,16 @@ def create_endpoints_dict(self) -> None:
+            # Email or id arguments include email and
+            # end with id so find and replace them with
+            # email_or_id_regex
+            path_regex = re.sub(r'{[^}]*email[^}]*id}', email_or_id_regex, path_regex)
```
What does this logic do?  I realize we have similar adjacent code but it's all pretty hacky and I'd prefer to not add more.

tushar912 replied:
This code was to match <code>email_or_id</code> param and replace it with <code> email_or_id</code> .I also feel that we can go without it as I don't find much use of it.

tushar912 replied:
I removed it .

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -4278,16 +4278,16 @@ paths:
           in: path
           description: |
-            The email address of the user whose presence you want to fetch.
+            The email address or user_id of the user whose presence you want to fetch.
```
We need to document the changes here; search for "Changes" in this file for an example, and document in `templates/zerver/api/changelog.md` as well.

tushar912 replied:
Ok

timabbott commented:
Thanks for working on this @tushar912! This is a pretty clean implementation if we're going to go with this model.  An alternative available to us is to use this URL syntax that we use elsewhere in `urls.py`:
```
    rest_path('user_groups/<int:user_group_id>',                                                     
```
for the integer case, and a second (string) copy for the string case.  And then we'd do the API documentation recording `{user_id}` as the value with a textual mention just in the documentation (not the URL scheme) that emails are supported as well?  

I suppose what model makes sense depends probably on whether we intend to eventually deprecate the email format, which I suppose we probably won't; in which case I think you've picked the right strategy.  But maybe we should migrate to use `user_id_or_email` as the variable name?  I think that expresses more clearly that the default is a user ID.

tushar912 commented:
@timabbott As per your suggestion I have replaced the name of param from <code>email_or_id</code> to <code>user_id_or_email</code> . I have also added the changes in <code>changelog.md</code>.

tushar912 commented:
@timabbott Can you review again.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -4227,16 +4227,19 @@ paths:
+            The user_id or email address of the user whose presence you want to fetch.
+
+            **Changes**: New in Zulip 4.0 (feature level 39). Previous versions supproter only
+            user_id .
```
Can you fix the typos here?  (`supported` mispelled, and an extra space before the `.` at the end of the Changes sentence).

For user-facing documentation like this, it's really important to have clear, copy-edited English, because typos really reduce confidence in the quality of documentation one's reading (I'm sure you've had this experience yourself).  So whenever editing a documentation file, please take special care to check for typos.

tushar912 replied:
Oh sorry I will take extra care next time.

timabbott commented on zerver/views/presence.py:
```diff
@@ -13,25 +13,45 @@
+        else:
+            target = get_active_user_profile_by_id_in_realm(
+                int(user_id_or_email), user_profile.realm
+            )
```
I wish we could make this simpler.  Would it read better to do:
```
try:
     user_id = int(user_id_or_email)
     target = get_active_user_profile_by_id_in_realm(user_id, user_profile.realm)
except ValueError:
     email = user_id_or_email
     target = get_active_user(user_id_or_email, user_profile.realm)
```

maybe all of that inside an outer `try/except` for `UserProfile.DoesNotExist`?

tushar912 replied:
Sure I would make the change.

timabbott commented on templates/zerver/api/changelog.md:
```diff
@@ -10,6 +10,11 @@ below features are supported.
+**Feature level 39**
+
+* [`GET /users/<user_id_or_email>/presence`]: Added support for user_id
+in `get_user_presence` endpoint.
```
Let's write this as "Added support for passing a `user_id` to identify the target user." -- no need to repeat the endpoint name via the less-official `get_user_presence` naming schema (which doesn't appear elsewhere in user-facing documentation)

Also, this needs a rebase an update to a new feature level, since we've merged other API migrations since you first wrote this.

Finally, a few of the details in your commit message, like "This is ready for review" and "Added changes to changelog" are unnecessary and would be better removed.

tushar912 replied:
Ok

timabbott commented:
@tushar912 thanks for working on this!  I think we're pretty close, but posted a batch of small things worth fixing.

tushar912 commented:
@timabbott I have made the changes.

timabbott commented:
This is great!  I merged this after the following changes:
* Rebased onto master, bumping ZULIP_FEATURE_LEVEL to 43 as the resolution to merge conflicts.
* Tweaked the reference to the endpoint in `changelog.md` to use `{` rather than `<`, since that's our pattern.
* Some very small edits in the API documentation copy.
* Edited the commit message to explain the background around the issue a bit more.
* Re-added the `test_openapi.py` test line that was deleted, after adjusting it for the new URL.  In the future, please comment as to why in the PR thread whenever deleting a test, since it's rarely the case that we want to do that as part of adding a new feature.

Thanks @tushar912!
